### PR TITLE
Expose StackTraceHiddenAttribute publicly

### LINKED ISSUE #29681: Consider exposing `System.Diagnostics.StackTraceHiddenAttribute` publicly
This attribute hides methods and types marked by it in the error message stacktraces. It is used extensively in the CoreLib to e.g. hide implementation details of async methods that are always the same and just add clutter to the error messages.

It appears to be a very useful attribute, particular as `System.ThrowHelper` is `internal`, this would be invaluable for user-written throw helper types to keep the best possible exception semantics while still having the advantages of a throw helper. It could be abused, but not really in a "breaking" way, just using it stupidly could make debugging a little harder, so unless there is something I am *really* missing I cannot see a strong case against making it public.

### Proposed API

```csharp
namespace System.Diagnostics
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method | AttributeTargets.Constructor | AttributeTargets.Struct, Inherited = false)]
    public sealed class StackTraceHiddenAttribute : Attribute
    {
        public StackTraceHiddenAttribute() { }
    }
}
```

### PR DESCRIPTION
**Pull request by stephentoub**
Fixes #29681

### CODE REVIEW

File: src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using Xunit;
+
+namespace System.Tests
+{
+    public static class StackTraceHiddenAttributeTests
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs:9]: Add a test to validate that `NoOptimization | NoInlining` method with this attribute is actually hidden in the stacktrace?

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using Xunit;
+
+namespace System.Tests
+{
+    public class StackTraceHiddenAttributeTests
```
Comment (jkotas) [inline src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs:10]:
```suggestion
    public class StackTraceHiddenAttributeTests
```
```suggestion
    public static class StackTraceHiddenAttributeTests
```

Comment (stephentoub) [reply src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs:10]: Is your preference for most/all test methods to be static?  We're very inconsistent on that, but I *think* (I'd need to actually measure, as I could be wrong) we tended towards instance methods.

Comment (jkotas) [reply src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs:10]: I do not have a preference. I looked at two other random test files and both of them were static. Feel free to ignore it.

Comment (stephentoub) [reply src/libraries/System.Runtime/tests/System/Diagnostics/StackTraceHiddenAttributeTests.cs:10]: Ok.  We lack consistency and guidance here.  I'd be happy for us to put some in place.  But let's do it separately.

**Pull request merged**